File: kubernetes-client-examples/javascript/create_deployment.js
const k8s = require('@kubernetes/client-node');

const kc = new k8s.KubeConfig();
kc.loadFromDefault();

// Get namespace from the current context in the kubeconfig
let namespace = 'default'; // fallback
const currentContext = kc.getCurrentContext();
const contextObj = kc.contexts.find(ctx => ctx.name === currentContext);
if (contextObj && contextObj.namespace) {
  namespace = contextObj.namespace;
}

const deploymentName = 'client-javascript-deployment';

const deploymentSpec = {
  metadata: {
    name: deploymentName,
    labels: {
      app: 'client-javascript-app'
    }
  },
  spec: {
    replicas: 2,
    selector: {
      matchLabels: {
        app: 'client-javascript-app'
      }
    },
    template: {
      metadata: {
        labels: {
          app: 'client-javascript-app'
        }
      },
      spec: {
        containers: [
          {
            name: 'nginx-container',
            image: 'nginx:latest',
            ports: [
              {
                containerPort: 80
              }
            ]
          }
        ]
      }
    }
  }
};

const main = async () => {
  try {
    const AppsV1Api = kc.makeApiClient(k8s.AppsV1Api);
    const createDeploymentResponse = await AppsV1Api.createNamespacedDeployment(namespace, deploymentSpec);
    console.log("Deployment %s created in namespace %s", createDeploymentResponse.body.metadata.name, namespace);
  } catch (err) {
    console.error(err.body || err);
  }
};

main();

File: kubernetes-client-examples/rust/src/bin/create_deployment.rs
use kube::Api;
use kube::Client;
use kube::config::Config;
use k8s_openapi::api::apps::v1::Deployment;
use kube::api::PostParams;
use k8s_openapi::api::core::v1::{Container, ContainerPort, PodSpec, PodTemplateSpec};
use k8s_openapi::apimachinery::pkg::apis::meta::v1::{ObjectMeta, LabelSelector};
use k8s_openapi::api::apps::v1::DeploymentSpec;

use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Load kubeconfig and get client
    let config = Config::infer().await?;
    let namespace = config.default_namespace.to_string();

    let client = Client::try_from(config)?;

    // Use the namespace from the kubeconfig context dynamically
    let deployments: Api<Deployment> = Api::namespaced(client.clone(), &namespace);

    let deployment_name = "nginx-deployment";
    let deployment = Deployment {
        metadata: ObjectMeta {
            name: Some(deployment_name.to_string()),
            labels: Some(
                vec![("app".to_string(), "nginx".to_string())]
                    .into_iter()
                    .collect(),
            ),
            ..Default::default()
        },
        spec: Some(DeploymentSpec {
            replicas: Some(3),
            selector: LabelSelector {
                match_labels: Some(
                    vec![("app".to_string(), "nginx".to_string())]
                        .into_iter()
                        .collect(),
                ),
                ..Default::default()
            },
            template: PodTemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(
                        vec![("app".to_string(), "nginx".to_string())]
                            .into_iter()
                            .collect(),
                    ),
                    ..Default::default()
                }),
                spec: Some(PodSpec {
                    containers: vec![
                        Container {
                            name: "nginx".to_string(),
                            image: Some("nginx:latest".to_string()),
                            ports: Some(vec![ContainerPort {
                                container_port: 80,
                                ..Default::default()
                            }]),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
            },
            ..Default::default()
        }),
        ..Default::default()
    };

    // Apply the deployment (create it)
    let post_params = PostParams::default();
    let result = deployments.create(&post_params, &deployment).await;

    match result {
        Ok(_) => println!("Deployment '{}' created successfully in namespace '{}'", deployment_name, namespace),
        Err(err) => eprintln!("Error creating deployment: {:?}", err),
    }

    Ok(())
}
